packages/gatsby-link/index.d.ts

```diff
@@ -1,11 +1,21 @@
-import * as React from "react";
+import React from "react"
+import { LinkProps } from "reach__router"
 
-export const push: (to: LocationDescriptor) => void;
-export const replace: (to: LocationDescriptor) => void;
+export const push: (to: string) => void
+export const replace: (to: string) => void
 
 // TODO: Remove navigateTo for Gatsby v3
-export const navigateTo: (to: LocationDescriptor) => void;
+export const navigateTo: (to: string) => void
 
-export const withPrefix: (path: string) => string;
+export const withPrefix: (path: string) => string
 
-export default class GatsbyLink extends React.Component<GatsbyLinkProps, any> { }
+export interface GatsbyLinkProps extends LinkProps<any> {
```

Comment:
We could add a generic here:

```ts
export interface GatsbyLinkProps<TState = {}> extends LinkProps<TState> {
  // ...
}
```

Also, some props (like `onClick`) are included within `LinkProps` (extended from `React.DetailedHTMLProps<React.AnchorHTMLAttributes<HTMLAnchorElement>, HTMLAnchorElement>` so they can be omitted.